### ISSUE TITLE
Do we need a built-in version of GLFW?

### ISSUE THREAD

**Issue by tsgouros** [ivlab/MinVR2#81]
MinVR built using its own version of GLFW interferes with a user's attempt to port an application built using the system-installed version of GLFW. Can we change so it only installs and links to its own version if there is no system version and otherwise uses the system version?

_objc[8085]: Class GLFWContentView is implemented in both yurt/minvr-images/build/bin/tgl and yurt/MinVR2-master/build/plugins/MinVR_GLFW/lib/libMinVR_GLFWd.dylib. One of the two will be used. Which one is undefined.
objc[8085]: Class GLFWWindow is implemented in both yurt/minvr-images/build/bin/tgl and yurt/MinVR2-master/build/plugins/MinVR_GLFW/lib/libMinVR_GLFWd.dylib. One of the two will be used. Which one is undefined.
objc[8085]: Class GLFWApplication is implemented in both yurt/minvr-images/build/bin/tgl and yurt/MinVR2-master/build/plugins/MinVR_GLFW/lib/libMinVR_GLFWd.dylib. One of the two will be used. Which one is undefined.
...
ABOUT TO INITIALIZE GLFW
libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: Failed to open GLFW window._ 

p.s. For reasons I don't understand, the MinVR version of GLFW does not work with outside-MinVR applications, and produces a ton of linker errors about symbols not found that belong to some OS X Framework:

  "_CFStringGetMaximumSizeForEncoding", referenced from:
      _getDisplayName in libglfw3.a(cocoa_monitor.m.o)
  "_CFURLCopyLastPathComponent", referenced from:
      _changeToResourcesDirectory in libglfw3.a(cocoa_init.m.o)
  "_CFURLGetFileSystemRepresentation", referenced from:
      _changeToResourcesDirectory in libglfw3.a(cocoa_init.m.o)
... etc etc etc etc.

**Comment by dtorban** [comment]
Yes, we can do exactly what you suggest.  Look for GLFW first then build it if it isn't found.  One reason that we use our own version of GLFW is because Bret created a branch that allows use of the GPU Affinity extension on Windows.  It allows us to pick which GPU we are using, but definitely this can be another plugin or cmake configuration itself.  I'll look into changing this.

**Comment by dtorban** [comment]
Also, to get rid of those errors for now, try linking your application with the cocoa libraries:

```
if (${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
	set(CMAKE_XCODE_ATTRIBUTE_CLANG_CXX_LANGUAGE_STANDARD "c++11")
	set(CMAKE_XCODE_ATTRIBUTE_CLANG_CXX_LIBRARY "libc++")
	find_library(COCOA_LIB Cocoa)
	find_library(IOKIT_LIB IOKit)
	find_library(CORE_FOUNDATION_FRAMEWORK CoreFoundation)
    	find_library(CORE_VIDEO_FRAMEWORK CoreVideo)
	set(LIBS_ALL ${LIBS_ALL} ${COCOA_LIB} ${IOKIT_LIB} ${CORE_FOUNDATION_FRAMEWORK} ${CORE_VIDEO_FRAMEWORK})
	message(STATUS "${CORE_VIDEO_FRAMEWORK}")
endif()
```

**Comment by tsgouros** [comment]
Thanks Dan. Do you have any idea why switching from one GLFW to another (from 3.1.2 to the one in the GLFW plugin) would invert the image and (maybe) reverse some coordinates?

**Comment by bretjackson** [comment]
The reason it uses its own version is that I modified GLFW to enable gpu
affinity, which was a pretty substantial performance improvement on the
Windows-based ivlab cave with multiple gpus (Otherwise, on Windows, all
opengl calls get sent to every gpu regardless of where the window is
rendering). I had an open pull request to integrate my changes into the
main GLFW for years, but the maintainer didn't really understand what it
was for and eventually closed it. It could be worth trying to get them
integrated again, but I won't waste my time trying again.

That being said, we should probably make it so that users can use their own
version of glfw if they want. The issue is that the VRGLFWWindowToolkit has
calls to glfwWindowHint to setup the gpu affinity. I don't know what the
base version of glfw will do if you pass in an enum value that doesn't
exist, but I would assume it is a compiler error. It's something to look
into.

The mac compiler errors might be fixed updating our version of glfw with
whatever is current. I'm not sure.

-Bret

**Comment by tsgouros** [comment]
I get it. I'll happily use that version of glfw if I can. But though it works within the context of a MinVR build, it doesn't work when an external program links to it. (Dan's suggestion of the IOKit, etc, frameworks appears to be exactly correct.)

Thanks. I'll leave this open until someone changes the build to either look for a locally-installed version (and if it finds one, maybe cmake should print a message suggesting an upgrade to Bret's version) or there's a flag to indicate which glfw to use.

**Comment by tsgouros** [comment]
So it turns out that the issue is that the GLFW included with the MinVR release has something screwed up about the function glfwGetCursorPos and possibly glfwSetCursorPos. I was using the cursor position to calculate a view matrix, which is why the object in my test program disappeared when I changed the GLFW version I linked to. I haven't spent the time to align the GLFW versions timeline with the bug reports google finds for me, but there have been, in the past, bugs reported about these cursor position functions.

The GLFW version we're including needs an upgrade, but I recommend we do not include any GLFW version in our release software. I get that there are improvements to GLFW that are important to people, but my suggestion is that these be included in the form of a README and a patch file for the GLFW release, or that there be a separate GLFW-fixed repo created, or something like that. Please close this issue when both the CMake issues and the GLFW issues are resolved.

**Comment by dtorban** [comment]
Okay, I finally understand what the problem is here.  The problem is not with the GLFW version that is built, the problem is that GLFW is being linked to from the application.  The whole purpose of MinVR  was to enable developers so they don't have to worry about linking or what versions of window toolkits they were using.  All this is abstracted out through the plugins and the data index.  Application developers should not link to GLFW because unintended consequences could definitely happen (as you have already experienced).  There could be versioning issues or different instances of the same objects created that don't reference each other.  Although their interfaces are the same, their memory footprints may be different.

Instead of writing a callback for glfwGetCursorPos, application developers should use the event returned in the event loop "/Mouse_Move" returned from the data index, which also has "/Mouse_Move/XPos" and "/Mouse_Move/YPos".  This is the same event that is returned if you are using GLUT instead of GLFW.  Similarly, instead of a callback for mouse presses, a developer should use "/Mouse_Down" or "/Mouse_Up" in the event loop.

Here is what the MinVR application code might look like:

```c++
virtual void onVREvent(const std::string &eventName, VRDataIndex *eventData) {
	//std::cout << "Event: " << eventName << std::endl;
	if (eventName == "/Mouse_Move") {
		double xpos = eventData->getValue("XPos", eventName);
		double ypos = eventData->getValue("YPos", eventName);
		std::cout << xpos << ", " << ypos << std::endl;
	}
}
```

Solution to this issue:

- Do not link to any GLFW version in the application code.
- Use the onVREvent to get window events like mouse moving, mouse presses, or keyboard events.
- Trust the plugin that it will handle everything related to GLFW (let it decide which version to use)

I think if all these are true, Bret's version of GLFW is just fine for now.  In fact, it might be prefered to tie to a version for consistency.  Otherwise, perhaps they should be two separate plugins: GLFW_Affinity and GLFW (I will log this as a separate issue).

**Comment by dtorban** [comment]
Added issue #82 to handle plugin versioning of GLFW.

**Comment by tsgouros** [comment]
Fair enough, and yes, it's unlikely that a VR application will use the mouse position.  But let's remember that part of the original design goal of MinVR was to allow developers to port their OpenGL programs to VR easily. I was converting a program that used GLFW, not expecting that my code would break just by linking with MinVR.

**Comment by tsgouros** [comment]
This is not fixed yet, but I'm closing this because #82 exists.
